### ISSUE TITLE
[RISC-V] Vector -flto -O2 miscompile

### ISSUE THREAD

**Issue by patrick-rivos** [llvm/llvm-project#86620]
Testcase:
```c
unsigned a, b;
signed char d = 0xBF;
long c[12];
int main() {
  c[9] = 0xFFFFF1FFF;
  for (int e = 0; e < 12; e += d - 0xFFFFFFBC) {
    a = c[e];
    b = ({ b > (unsigned short)~c[e] ? b : (unsigned short)~c[e]; });
  }
  __builtin_printf("%X\n", a);
}
```

Commands:
```bash
> /scratch/tc-testing/tc-mar-25/build-rv64gcv/bin/clang -march=rv64gcv -flto -O2 -fuse-ld=lld red.c -o red.out -fwrapv
> /scratch/tc-testing/tc-mar-25/build-rv64gcv/bin/qemu-riscv64 red.out
1FFF
> /scratch/tc-testing/tc-mar-25/build-rv64gcv/bin/clang red.c -o red.out -fwrapv
> /scratch/tc-testing/tc-mar-25/build-rv64gcv/bin/qemu-riscv64 red.out
FFFF1FFF
```

Reduced LLVM IR:
```llvm ir
target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128"
target triple = "riscv64-unknown-linux-gnu"

@d = global i8 -65
@c = global [12 x i64] zeroinitializer
@b = global i32 0
@.str = constant [4 x i8] c"%X\0A\00"

define i32 @main() #0 {
entry:
  store i64 68719419391, ptr getelementptr inbounds ([12 x i64], ptr @c, i64 0, i64 9), align 8
  %0 = load i8, ptr @d, align 1
  %conv12 = sext i8 %0 to i32
  %b.promoted = load i32, ptr @b, align 4
  br label %for.body

for.cond.cleanup:                                 ; preds = %for.body
  store i32 %.conv4, ptr @b, align 4
  %call = tail call i32 (ptr, ...) @printf(ptr @.str, i32 %conv)
  ret i32 0

for.body:                                         ; preds = %for.body, %entry
  %e.020 = phi i32 [ 0, %entry ], [ %add, %for.body ]
  %.conv41819 = phi i32 [ %b.promoted, %entry ], [ %.conv4, %for.body ]
  %idxprom = sext i32 %e.020 to i64
  %arrayidx = getelementptr [12 x i64], ptr @c, i64 0, i64 %idxprom
  %1 = load i64, ptr %arrayidx, align 8
  %conv = trunc i64 %1 to i32
  %conv3 = and i32 %conv, 1
  %conv4 = xor i32 %conv3, 65535
  %.conv4 = tail call i32 @llvm.umax.i32(i32 %.conv41819, i32 %conv4)
  %sub = add i32 %e.020, 68
  %add = add i32 %sub, %conv12
  %cmp = icmp slt i32 %add, 12
  br i1 %cmp, label %for.body, label %for.cond.cleanup

; uselistorder directives
  uselistorder i32 %.conv4, { 1, 0 }
}

declare i32 @printf(ptr, ...)

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare i32 @llvm.umax.i32(i32, i32) #1

attributes #0 = { "target-features"="+64bit,+v" }
attributes #1 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }

!llvm.module.flags = !{!0}

!0 = !{i32 1, !"target-abi", !"lp64d"}
```

Commands:
```bash
> /scratch/tc-testing/tc-mar-25/build-rv64gcv/bin/clang -march=rv64gcv -O2 reduced.ll -o red.out -fwrapv
> /scratch/tc-testing/tc-mar-25/build-rv64gcv/bin/qemu-riscv64 red.out
FFFF1FFF

> /scratch/tc-testing/tc-mar-25/build-rv64gcv/bin/clang -march=rv64gcv -O2 -flto -fuse-ld=lld reduced.ll -o red.out -fwrapv
> /scratch/tc-testing/tc-mar-25/build-rv64gcv/bin/qemu-riscv64 red.out
1FFF
```

Found via fuzzer.

**Comment by topperc** [comment]
I think this may be incorrect SLP vectorization

```
  define dso_local noundef signext i32 @main() local_unnamed_addr #0 {             
  entry:                                                                           
    store i64 68719419391, ptr getelementptr inbounds ([12 x i64], ptr @c, i64 0, i64 9), align 8, !tbaa !11
    %b.promoted = load i32, ptr @b, align 4, !tbaa !15                             
    %0 = load i64, ptr @c, align 8, !tbaa !11                                      
    %conv = trunc i64 %0 to i32                                                    
    %conv3 = and i32 %conv, 65535                                                  
    %conv4 = xor i32 %conv3, 65535                                                 
    %.conv4 = tail call i32 @llvm.umax.i32(i32 %b.promoted, i32 %conv4)            
    %1 = load i64, ptr getelementptr inbounds ([12 x i64], ptr @c, i64 0, i64 3), align 8, !tbaa !11
    %conv.1 = trunc i64 %1 to i32                                                  
    %conv3.1 = and i32 %conv.1, 65535                                              
    %conv4.1 = xor i32 %conv3.1, 65535                                             
    %.conv4.1 = tail call i32 @llvm.umax.i32(i32 %.conv4, i32 %conv4.1)            
    %2 = load i64, ptr getelementptr inbounds ([12 x i64], ptr @c, i64 0, i64 6), align 8, !tbaa !11
    %conv.2 = trunc i64 %2 to i32                                                  
    %conv3.2 = and i32 %conv.2, 65535                                              
    %conv4.2 = xor i32 %conv3.2, 65535                                             
    %.conv4.2 = tail call i32 @llvm.umax.i32(i32 %.conv4.1, i32 %conv4.2)          
    %3 = load i64, ptr getelementptr inbounds ([12 x i64], ptr @c, i64 0, i64 9), align 8, !tbaa !11  
    %conv.3 = trunc i64 %3 to i32                                                                     
    %conv3.3 = and i32 %conv.3, 65535                                                                 
    %conv4.3 = xor i32 %conv3.3, 65535                                                                
    %.conv4.3 = tail call i32 @llvm.umax.i32(i32 %.conv4.2, i32 %conv4.3)          
    store i32 %.conv4.3, ptr @b, align 4, !tbaa !15                                                   
    %call = tail call signext i32 (ptr, ...) @printf(ptr noundef nonnull dereferenceable(1) @.str, i32 noundef signext %conv.3) #3
    ret i32 0                                                                                         
  } 
```

becomes

```
  define dso_local noundef signext i32 @main() local_unnamed_addr #0 {             
  entry:                                                                           
    store i64 68719419391, ptr getelementptr inbounds ([12 x i64], ptr @c, i64 0, i64 9), align 8, !tbaa !11
    %b.promoted = load i32, ptr @b, align 4, !tbaa !15                             
    %0 = call <4 x i64> @llvm.experimental.vp.strided.load.v4i64.p0.i64(ptr align 8 @c, i64 24, <4 x i1> <i1 true, i1 true, i1 true, i1 true>, i32 4), !tbaa !11
    %1 = trunc <4 x i64> %0 to <4 x i16>                                           
    %2 = and <4 x i16> %1, <i16 -1, i16 -1, i16 -1, i16 -1>                        
    %3 = xor <4 x i16> %2, <i16 -1, i16 -1, i16 -1, i16 -1>                        
    %4 = zext <4 x i16> %3 to <4 x i32>                                            
    %5 = call i32 @llvm.vector.reduce.umax.v4i32(<4 x i32> %4)                     
    %6 = call i32 @llvm.umax.i32(i32 %5, i32 %b.promoted)                          
    store i32 %6, ptr @b, align 4, !tbaa !15                                       
    %7 = extractelement <4 x i16> %1, i32 3                                        
    %8 = sext i16 %7 to i32                                                        
    %call = tail call signext i32 (ptr, ...) @printf(ptr noundef nonnull dereferenceable(1) @.str, i32 noundef signext %8) #4
    ret i32 0                                                                      
  } 
```

The printed value should be bits 31:0 of what the strided load got for element 3, but instead it was truncated to i16 and then sign extended back to i32.

**Comment by alexey-bataev** [comment]
Fixed in 342f7d0d35774ffdcb56e8b92252763a59bd2c29
